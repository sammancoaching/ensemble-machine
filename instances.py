#!/usr/bin/env python3
import datetime
import json
from dataclasses import dataclass
from pathlib import Path

import boto3
import click

from summon import read_regions_config

@dataclass
class RunningInstance:
    ip_address: str
    state: str
    coach: str
    url: str
    creation_date: str


@click.command()
@click.option(
    "--region-name",
    help=f"the aws region name (default eu-north-1): {', '.join(read_regions_config().keys())}",
    default="eu-north-1"
)
@click.option(
    "--aws-profile",
    default=None,
    help="the aws profile, if you dont use the default"
)
@click.option(
    "--coach",
    help="only show instances owned by this person",
)
def main(region_name, aws_profile, coach=None):
    instances = all_instances(region_name, aws_profile)
    if coach:
        instances = [instance for instance in instances if instance.coach.strip() == coach]
    print('\n'.join(print_instances(instances)))


def all_instances(region_name, aws_profile):
    response = get_sammancoach_machines(region_name, aws_profile)
    instances = instances_from_response(response)
    return instances


def print_instances(instances):
    rows = []
    for instance in instances:
        rows.append(print_instance(instance))
    return rows


def get_sammancoach_machines(region_name, aws_profile):
    session = boto3.Session(profile_name=aws_profile, region_name=region_name)
    client = session.client('ec2')
    custom_filter = [{
        'Name': 'tag:SammanCoach',
        'Values': ['*']}]
    response = client.describe_instances(Filters=custom_filter)
    return response


def instances_from_response(obj):
    machines = []
    for reservation in obj['Reservations']:
        for instance in reservation['Instances']:
            tags = instance['Tags']
            public_ip = instance.get("PublicIpAddress", "not available")
            attach_time = try_find_attach_time(instance)
            attach_time_short = attach_time.partition('T')[0] if attach_time else '-'
            machine_data = {
                "State": instance["State"]["Name"],
                "IP": public_ip,
                "AttachTime": attach_time_short,
            }
            for machine in tags:
                key = machine['Key']
                value = machine['Value']
                machine_data[key] = value
            machines.append(machine_data)
    instances = []
    for machine in machines:
        # TODO: find better way to filter out projector machines!
        if 'codekata.proagile.link' in machine['Name']:
            instance = RunningInstance(ip_address=f"{machine['IP']:16}",
                                       state=f"{machine['State']:12}",
                                       coach=f"{machine['SammanCoach']:12}",
                                       url=f"https://{machine['Name']:42}",
                                       creation_date=f"{machine['AttachTime']}")
            instances.append(instance)
    return instances


def print_instance(instance: RunningInstance):
    return f"{instance.ip_address} {instance.state} {instance.coach} {instance.url} {instance.creation_date}"


def try_find_attach_time(instance):
    try:
        time_value = instance["BlockDeviceMappings"][0]["Ebs"]["AttachTime"]
        if isinstance(time_value, datetime.datetime):
            return time_value.isoformat()
        else:
            return time_value
    except:
        return None


if __name__ == '__main__':
    import logging

    logging.basicConfig(level=logging.INFO)
    main()
