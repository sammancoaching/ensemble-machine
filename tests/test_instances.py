import datetime
import json
from pathlib import Path

from approvaltests import verify, verify_all

from instances import get_sammancoach_machines, RunningInstance, print_instance, \
    instances_from_response

SAMPLE_RESPONSE = Path('tests/response.json')


def test_instance_list_report():
    # To update the sample response (boto3/ec2 API changes)
    # run this test as a script in repo root:
    #     PYTHONPATH=. python tests/test_instances.py
    # That will overwrite the sample response file with a fresh response
    sample_response = SAMPLE_RESPONSE.read_text(encoding='utf8')
    output = instance_report_from_response(sample_response)
    verify_all('instances', output)


def datetime_handler(x):
    if isinstance(x, datetime.datetime):
        return x.isoformat()
    raise TypeError("Unknown type")


def instance_report_from_response(json_response):
    obj = json.loads(json_response)
    return instances_from_response(obj, "codekata.proagile.link")


def save_new_response_json():
    response_obj = get_sammancoach_machines("eu-north-1", None)
    response_json = json.dumps(response_obj, default=datetime_handler)
    SAMPLE_RESPONSE.write_text(response_json, encoding='utf8')


def test_print_instance():
    instance = RunningInstance("18.157.73.25", "running", "emily", "99daacf1-rider.codekata.proagile.link", "2022-01-31", region_name="eu-north-1")
    verify(print_instance(instance))

if __name__ == '__main__':
    save_new_response_json()
