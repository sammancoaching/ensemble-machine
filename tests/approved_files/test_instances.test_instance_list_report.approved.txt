instances

0) RunningInstance(ip_address='not available   ', state='stopped     ', coach='emily       ', url='https://3a73353f-clion.codekata.proagile.link     ', creation_date='2021-05-31', region_name='eu-north-1c')
1) RunningInstance(ip_address='not available   ', state='stopped     ', coach='ebjaolo     ', url='https://692fe0c2-rider.codekata.proagile.link     ', creation_date='2021-06-03', region_name='eu-north-1c')
2) RunningInstance(ip_address='not available   ', state='stopped     ', coach='ebjaolo     ', url='https://1d2ff333-clion.codekata.proagile.link     ', creation_date='2021-06-11', region_name='eu-north-1c')
3) RunningInstance(ip_address='not available   ', state='stopped     ', coach='ebjaolo     ', url='https://0899b6a8-clion.codekata.proagile.link     ', creation_date='2021-05-06', region_name='eu-north-1a')
4) RunningInstance(ip_address='not available   ', state='stopped     ', coach='emily       ', url='https://808c807c-pycharm.codekata.proagile.link   ', creation_date='2021-05-20', region_name='eu-north-1a')
5) RunningInstance(ip_address='not available   ', state='stopped     ', coach='ebjaolo     ', url='https://26c62aa1-clion.codekata.proagile.link     ', creation_date='2021-05-25', region_name='eu-north-1a')
6) RunningInstance(ip_address='not available   ', state='stopped     ', coach='emily       ', url='https://41e8a61e-webstorm.codekata.proagile.link  ', creation_date='2021-07-05', region_name='eu-north-1a')
7) RunningInstance(ip_address='not available   ', state='stopped     ', coach='emily       ', url='https://b5fdbf2c-webstorm.codekata.proagile.link  ', creation_date='2021-08-18', region_name='eu-north-1a')
8) RunningInstance(ip_address='not available   ', state='stopped     ', coach='emily       ', url='https://f8a3f1e4-webstorm.codekata.proagile.link  ', creation_date='2021-08-19', region_name='eu-north-1a')
9) RunningInstance(ip_address='not available   ', state='stopped     ', coach='emily       ', url='https://8d64b7cb-webstorm.codekata.proagile.link  ', creation_date='2021-08-19', region_name='eu-north-1a')
10) RunningInstance(ip_address='not available   ', state='stopped     ', coach='ebjaolo     ', url='https://abc5b7bc-clion.codekata.proagile.link     ', creation_date='2021-08-23', region_name='eu-north-1a')
11) RunningInstance(ip_address='not available   ', state='stopped     ', coach='ebjaolo     ', url='https://e569d3b1-idea.codekata.proagile.link      ', creation_date='2021-08-23', region_name='eu-north-1a')
12) RunningInstance(ip_address='not available   ', state='stopped     ', coach='emily       ', url='https://dbae4de0-idea.codekata.proagile.link      ', creation_date='2021-08-23', region_name='eu-north-1a')
13) RunningInstance(ip_address='not available   ', state='stopped     ', coach='ebjaolo     ', url='https://fad37321-idea.codekata.proagile.link      ', creation_date='2021-08-23', region_name='eu-north-1a')
14) RunningInstance(ip_address='not available   ', state='stopped     ', coach='ebjaolo     ', url='https://989f8e56-pycharm.codekata.proagile.link   ', creation_date='2021-08-24', region_name='eu-north-1a')
15) RunningInstance(ip_address='not available   ', state='stopped     ', coach='emily       ', url='https://a85d60be-rider-core.codekata.proagile.link', creation_date='2021-09-14', region_name='eu-north-1a')
16) RunningInstance(ip_address='not available   ', state='stopped     ', coach='emily       ', url='https://30432a29-idea.codekata.proagile.link      ', creation_date='2021-09-14', region_name='eu-north-1a')
17) RunningInstance(ip_address='not available   ', state='stopped     ', coach='neppord     ', url='https://76c39921-idea.codekata.proagile.link      ', creation_date='2021-07-07', region_name='eu-north-1c')
18) RunningInstance(ip_address='not available   ', state='stopped     ', coach='neppord     ', url='https://14c8ddae-pycharm.codekata.proagile.link   ', creation_date='2021-08-30', region_name='eu-north-1c')
19) RunningInstance(ip_address='not available   ', state='stopped     ', coach='emily       ', url='https://717ea5e2-erlang.codekata.proagile.link    ', creation_date='2021-08-31', region_name='eu-north-1c')
